tests/frontend/test_circuitspecs_chip2.py

```diff
@@ -0,0 +1,373 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Unit tests for the CircuitSpec class"""
+import textwrap
+
+import pytest
+import numpy as np
+import networkx as nx
+
+import blackbird
+
+import strawberryfields as sf
+import strawberryfields.ops as ops
+
+from strawberryfields.parameters import par_evaluate
+from strawberryfields.program_utils import CircuitError, list_to_DAG
+from strawberryfields.io import to_program
+from strawberryfields.utils import random_interferometer
+from strawberryfields.circuitspecs.chip2 import Chip2Specs, CircuitSpecs
+
+
+pytestmark = pytest.mark.frontend
+
+np.random.seed(42)
+
+
+def program_equivalence(prog1, prog2, compare_params=True, atol=1e-6, rtol=0):
+    r"""Checks if two programs are equivalent.
+
+    This function converts the program lists into directed acyclic graphs,
+    and runs the NetworkX `is_isomorphic` graph function in order
+    to determine if the two programs are equivalent.
+
+    Note: when checking for parameter equality between two parameters
+    :math:`a` and :math:`b`, we use the following formula:
+
+    .. math:: |a - b| \leq (\texttt{atol} + \texttt{rtol}\times|b|)
+
+    Args:
+        prog1 (strawberryfields.program.Program): quantum program
+        prog2 (strawberryfields.program.Program): quantum program
+        compare_params (bool): Set to ``False`` to turn of comparing
+            program parameters; equivalency will only take into
+            account the operation order.
+        atol (float): the absolute tolerance parameter for checking
+            quantum operation parameter equality
+        rtol (float): the relative tolerance parameter for checking
+            quantum operation parameter equality
+
+    Returns:
+        bool: returns ``True`` if two quantum programs are equivalent
+    """
+    DAG1 = list_to_DAG(prog1.circuit)
+    DAG2 = list_to_DAG(prog2.circuit)
+
+    circuit = []
+    for G in [DAG1, DAG2]:
+        # relabel the DAG nodes to integers
+        circuit.append(nx.convert_node_labels_to_integers(G))
+
+        # add node attributes to store the operation name and parameters
+        name_mapping = {i: n.op.__class__.__name__ for i, n in enumerate(G.nodes())}
+        parameter_mapping = {i: par_evaluate(n.op.p) for i, n in enumerate(G.nodes())}
+
+        # CXgate and BSgate are not symmetric wrt to permuting the order of the two
+        # modes it acts on; i.e., the order of the wires matter
+        wire_mapping = {}
+        for i, n in enumerate(G.nodes()):
+            if n.op.__class__.__name__ == "CXgate":
+                if np.allclose(n.op.p[0], 0):
+                    # if the CXgate parameter is 0, wire order doesn't matter
+                    wire_mapping[i] = 0
```

Comment:
is `0` a placeholder/indicator? the two branches of the conditional seem to lead to different types in `wire_mapping[i]`

Reply:
Let me think, it's been a while since I wrote this. I suppose the thinking is that `CXgate(0) | [2, 1]` should be equivalent to `CXgate(0) | [1, 2]`, but `CXgate(0.5) | [1, 2]` and `CXgate(0.5) | [2, 1]` must be treated as distinct gates.

Reply:
I understood that part. I was just asking about why use `0` if the gate clearly has two wires?

Reply:
I don't remember :laughing: Programmatic shortcut? Probably a placeholder like you mentioned.